File: BFS & DFS/programmers43163.cpp
// 문제 및 풀이: https://velog.io/@e7838752/programmers43163
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    bool visited[50] = {0, };
    
    queue<pair<string, int>> q;
    q.push({begin, 0});
    
    while (!q.empty()) {
        string word = q.front().first;
        int depth = q.front().second;
        q.pop();
        
        if (word == target) {
            answer = depth;
            break;
        }
        
        for (int i = 0; i < words.size(); ++i) {
            if (visited[i]) continue;
            int count = 0;
            
            for (int j = 0; j < words[i].size(); ++j) {
                if (word[j] == words[i][j]) count++;
            }
            
            if (count == words[i].size() - 1) {
                q.push({words[i], depth + 1});
                visited[i] = true;
            }
        }
        
    }
    
    return answer;
}

File: BFS_and_DFS/programmers43163.js
function solution(begin, target, words) {
  let answer = 10e9;
  let visited = Array(words.length).fill(false);

  function dfs(word, cnt) {
    if (word === target) {
      answer = Math.min(answer, cnt);
    } else {
      for (let i = 0; i < words.length; i++) {
        if (visited[i]) continue;

        let match = 0;
        for (let j = 0; j < words[i].length; j++) {
          if (words[i][j] === word[j]) match++;
        }

        if (match === word.length - 1) {
          visited[i] = true;
          dfs(words[i], cnt + 1);
          visited[i] = false;
        }
      }
    }
  }

  dfs(begin, 0);
  answer = answer === 10e9 ? 0 : answer;
  return answer;
}
